mantis-publish/mantis-publish-core/src/main/java/io/mantisrx/publish/AbstractSubscriptionTracker.java

```diff
@@ -66,52 +72,51 @@ public AbstractSubscriptionTracker(MrePublishConfiguration mrePublishConfigurati
-                }
-            } catch (Throwable t) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("failed to remove subscription {}", subToRemove, t);
+    void propagateSubscriptionChanges(String streamName, Set<MantisServerSubscription> curr) {
+        Set<String> previousIds = getCurrentSubIds(streamName);
+
+        for (MantisServerSubscription newSub : curr) {
+            // Add new subscription not present previously
+            if (!previousIds.contains(newSub.getSubscriptionId())) {
+                try {
+                    Optional<Subscription> subscription = SubscriptionFactory.getSubscription(newSub.getSubscriptionId(), newSub.getQuery());
+                    if (subscription.isPresent()) {
+                        streamManager.addStreamSubscription(subscription.get());
+                    } else {
+                        LOG.info("will not add invalid subscription {}", newSub);
+                    }
+                } catch (Throwable t) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("failed to add subscription {}", newSub, t);
+                    }
                 }
             }
-        });
 
-        Set<MantisServerSubscription> currSubsNotInPrev = new HashSet<>(curr);
-        currSubsNotInPrev.removeAll(prev);
-        currSubsNotInPrev.stream().forEach(subToAdd -> {
+            previousIds.remove(newSub.getSubscriptionId());
+        }
+
+        // Remove previous subscriptions no longer present.
+        previousIds.stream().forEach(prevId -> {
             try {
-                Optional<Subscription> subscription = SubscriptionFactory.getSubscription(subToAdd.getSubscriptionId(), subToAdd.getQuery());
-                if (subscription.isPresent()) {
-                    streamManager.addStreamSubscription(subscription.get());
-                } else {
-                    LOG.info("will not add invalid subscription {}", subToAdd);
-                }
+                streamManager.removeStreamSubscription(prevId);
             } catch (Throwable t) {
                 if (LOG.isDebugEnabled()) {
-                    LOG.debug("failed to add subscription {}", subToAdd, t);
+                    LOG.debug("failed to remove subscription {}", prevId);
                 }
             }
         });
     }
 
 
     private void cleanupStaleSubscriptions(String streamName) {
-        StreamSubscriptions streamSubscriptions = previousSubscriptions.get(streamName);
-        if (streamSubscriptions != null) {
-            boolean hasStaleSubscriptionsData = (System.currentTimeMillis() - streamSubscriptions.getCreateTimeMs()) > mrePublishConfiguration.subscriptionExpiryIntervalSec() * 1000;
+        Long lastFetched = streamLastFetchedTs.get(streamName);
+        if (lastFetched != null) {
+            boolean hasStaleSubscriptionsData = (System.currentTimeMillis() - lastFetched) > mrePublishConfiguration.subscriptionExpiryIntervalSec() * 1000;
```

Comment:
might be a good idea to get rid of the dependency on the System clock here at a later point.

Reply:
What's the issue with using system clock for expiration? What would you replace it with?

Reply:
Because it makes it harder to test. Generally, I recommend something like 

https://github.com/Netflix/spectator/blob/606f3241057a4cd6fc0632625bd4ea11c5c65bc1/spectator-api/src/main/java/com/netflix/spectator/api/Clock.java 
Usage: https://github.com/Netflix/spectator/blob/606f3241057a4cd6fc0632625bd4ea11c5c65bc1/spectator-api/src/main/java/com/netflix/spectator/api/AbstractRegistry.java#L44

or just depending on java.time.Clock. 

Again, this is not required for this change as you didn't introduce it. just might be a good idea to replace it at some point IMO.